File: mpi/c/reduce.c
#include <stdio.h>
#include <stdlib.h>
#include "mpi.h"

int main( int argc, char *argv[] )
{
    int    i, nsize,
           nprocs, myrank, root ;
    double *array, *array_sum ;

    MPI_Init( &argc, &argv ) ;
    MPI_Comm_size( MPI_COMM_WORLD, &nprocs ) ;
    MPI_Comm_rank( MPI_COMM_WORLD, &myrank ) ;

    root = 0 ;
    nsize = 10000 ;
    array     = ( double * )malloc( nsize * sizeof( double ) ) ;
    for ( i = 0 ; i < nsize ; i++ ) {
	    array[ i ] = 3.938e-04 + ( i + 1 ) * 4 * ( myrank + 1 ) ;
    }

    if ( myrank == root ) {
        array_sum = ( double * )malloc( nsize * sizeof( double ) ) ;
        printf( "Nproc = %10d\n", nprocs ) ;
        printf( "Before reduce:\n" ) ;
        printf( "%16.8e, %16.8e\n", array[ 0 ], array[ nsize - 1 ] ) ;
    }

    MPI_Reduce( array, array_sum, nsize, MPI_DOUBLE, MPI_SUM, root, MPI_COMM_WORLD ) ;

    if ( myrank == root ) {
        printf( "Nproc = %10d\n", nprocs ) ;
        printf( "After  reduce:\n" ) ;
        printf( "%16.8e, %16.8e\n", array_sum[ 0 ], array_sum[ nsize - 1 ] ) ;
        
    }
    
    free( array ) ;
    if ( myrank == root ) {
        free( array_sum ) ;
    }
    MPI_Finalize() ;
    return 0 ;
}

File: mpi/python/reduce.py
import sys, numpy
from mpi4py import MPI

##########################################
##########################################
##########################################
comm   = MPI.COMM_WORLD
myrank = MPI.COMM_WORLD.Get_rank()
nprocs = MPI.COMM_WORLD.Get_size()
root   = 0

##########################################
##########################################
##########################################
nsize = 10000
array = numpy.empty( [ nsize ], numpy.float64 )
for i in range( nsize ):
	array[ i ] = 3.938e-04 + ( i + 1 ) * 4 * ( myrank + 1 )

##########################################
##########################################
##########################################
if ( myrank == root ):
    print( "Before reduce:" )
    print( "Nproc, rank = %10d%10d" % ( nprocs, myrank ) )
    print( "%16.8e, %16.8e" % ( array[ 0 ], array[ nsize - 1 ] ) )

##########################################
##########################################
##########################################
garray = comm.reduce( array, MPI.SUM, root )

if ( myrank == root ):
    print( "After  reduce:" )
    print( "Nproc, rank = %10d%10d" % ( nprocs, myrank ) )
    print( "%16.8e, %16.8e" % ( garray[ 0 ], garray[ nsize - 1 ] ) )
